### ISSUE TITLE
Terraform CCloud provider JWT authentication

### ISSUE THREAD

**Issue by ajfc84** [confluentinc/terraform-provider-confluent#240]
I'm trying to set up Gitlab's injected JWT tokens(id tokens) to set up authentication with Confluent Cloud. There is an OAuth endpoint present in Confluent that together with and Identity Provider resource set up with Gitlab's JWKS could be used as an authentication mechanism beyond the current API_Key mechanism present on the Terraform provider.
I suppose the only thing needed is to extend the existing Basic authentication header mechanism to support  Bearer header authentication?! Also to note is that things like the ACL resource still imply the use of an Api_Key.

Confluent Cloud already supports OAuth/STS authentication so can the Terraform provider also be updated to support token based authentication?!

**Comment by Nantero1** [comment]
We are also waiting for this feature! Would be nice. Any road map on this?

**Comment by linouk23** [comment]
Thanks for asking! It's on the roadmap 👌

**Comment by linouk23** [comment]
@ajfc84 @Nantero1 @S1M0NM @cbcoutinho @dylanmei @simon-an @lleiding 

Could you confirm that in other words you're essentially asking for a new supported auth method for TF Provider for Confluent like:
```
provider "confluent" {
  oauth_client_id = "..."
  oauth_client_secret = "..."
  oauth_endpoint = "..."
  oauth_identity_pool = "..." (required for now but could be optional in the future)
  oauth_identity_scope = "..." (optional -- might be necessary for OKTA)
}
```

so you could use it instead of Cloud API Keys / Kafka API Keys? Do these new input arguments make sense to you?

**Comment by Nantero1** [comment]
> @ajfc84 @Nantero1 @S1M0NM @cbcoutinho @dylanmei @simon-an @lleiding
> 
> Could you confirm that in other words you're essentially asking for a new supported auth method for TF Provider for Confluent like:
> 
> ```
> provider "confluent" {
>   oauth_client_id = "..."
>   oauth_client_secret = "..."
>   oauth_endpoint = "..." (double check whether it could be inferred from identity pool)
>   oauth_identity_pool = "..." (required for now but could be optional in the future)
>   oauth_identity_scope = "..." (optional -- might be necessary for OKTA)
> }
> ```
> 
> so you could use it instead of Cloud API Keys / Kafka API Keys? Do these new input arguments make sense to you?

Yes.

**Comment by dylanmei** [comment]
> so you could use it instead of Cloud API Keys / Kafka API Keys? Do these new input arguments make sense to you?

@linouk23 Yes. Identity pool is currently necessary when interacting with the Kafka API, and for converting OAuth token for an STS token, for example when listing environments or working with role bindings. In my setup, I always need to supply a scope.

```hcl
provider "confluent" {
  oauth_client_id      = var.oauth_client_id
  oauth_client_secret  = var.oauth_client_secret
  oauth_token_endpoint = var.oauth_token_endpoint
  oauth_scope          = var.oauth_scopes
  identity_pool_id     = var.identity_pool_id
}
```

**Comment by linouk23** [comment]
@dylanmei I've got a quick question: is it a single scope or multiple ones TF provider should accept (`oauth_scope: string` vs `oauth_scopes: [string]`? Thanks!

**Comment by dylanmei** [comment]
Scope is a list of values, but generally presented as a delimited string: [reference](https://help.okta.com/wf/en-us/content/topics/workflows/connector-builder/authentication-oauth2.htm)

 So it may contain several scopes, even though the singular parameter name may suggest otherwise.  [example](https://api.slack.com/legacy/oauth-scopes#working-with-scopes), [example](https://docs.confluent.io/cloud/current/access-management/authenticate/oauth/overview.html).

I don't think it's useful to use an Array type in the provider declaration.

**Comment by simon-an** [comment]
> @ajfc84 @Nantero1 @S1M0NM @cbcoutinho @dylanmei @simon-an @lleiding
> 
> Could you confirm that in other words you're essentially asking for a new supported auth method for TF Provider for Confluent like:
> 
> ```
> provider "confluent" {
>   oauth_client_id = "..."
>   oauth_client_secret = "..."
>   oauth_endpoint = "..." (double check whether it could be inferred from identity pool)
>   oauth_identity_pool = "..." (required for now but could be optional in the future)
>   oauth_identity_scope = "..." (optional -- might be necessary for OKTA)
> }
> ```
> 
> so you could use it instead of Cloud API Keys / Kafka API Keys? Do these new input arguments make sense to you?

Yes and no. Instead of oauth_client_secret = "..." i would need oauth_token, so any token generated by an idp and is valid for the oauth_identity_pool  could be used.

**Comment by linouk23** [comment]
@simon-an thanks for the message! Could you share your preferred definition of `provider "confluent" {}`? Thanks!

**Comment by simon-an** [comment]
`provider "confluent" {
  oauth_token = "..."
  oauth_identity_pool = "..." (required for now but could be optional in the future)
}`

So basically i want to provider to be able to this:


```yaml
BEARER_TOKEN=$(az account get-access-token | jq -r '.accessToken')
curl -S -v -X POST -H "Content-Type: application/json" -H "Confluent-Identity-Pool-Id: $POOL_ID" -H "Authorization: Bearer $BEARER_TOKEN" $CLUSTER_URL/kafka/v3/clusters/$CLUSTER_ID/topics -d '{"topic_name":"test"}')
```

**Comment by simon-an** [comment]
I would also like to see the oauth flow in  [client credential flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential
) with client_assertion and client_assertion_type instead of client_secret being implemented. This would allow to do:
CI-Server Token from IDP-1 -> IDP-2 with Federated Identity -> Confluent Cloud.
But this is optinal, because i can achive the same thing with the provider config above + 2 extra lines of script exchanging the token before doing any terraform commands.

**Comment by Nantero1** [comment]
> ```
> provider "confluent" {
>   oauth_token = "..."
>   oauth_identity_pool = "..." (required for now but could be optional in the future)
> }
> ```
> 
> So basically i want the provider to be able to this:
> 
> ```yaml
> BEARER_TOKEN=$(az account get-access-token | jq -r '.accessToken')
> curl -S -v -X POST -H "Content-Type: application/json" -H "Confluent-Identity-Pool-Id: $POOL_ID" -H "Authorization: Bearer $BEARER_TOKEN" $CLUSTER_URL/kafka/v3/clusters/$CLUSTER_ID/topics -d '{"topic_name":"test"}')
> ```
> 
> Which would allow any flow to be used and especially to reuse tokens that already exist in the current environment (like CI Server or Azure CLI in this example).

This confuses me. You pass in a token as input, but want a token to be retrieved at the same time by the provider? In my opinion this contradicts itself.

**Comment by matios13** [comment]
Hi, is there any plan to implement that feature?
